includes/sanitizers/class-amp-meta-sanitizer.php

```diff
@@ -0,0 +1,333 @@
+	 *
+	 * @return string|false Detected charset of the document, or false if none.
+	 */
+	protected function detect_charset() {
+		$charset = false;
+
+		// Check for HTML 4 http-equiv meta tags.
+		$http_equiv_tag = $this->xpath->query( '//meta[ @http-equiv and @content ]' )->item( 0 );
+		if ( $http_equiv_tag ) {
+			$http_equiv_tag->parentNode->removeChild( $http_equiv_tag );
+
+			// Check for the existence of a proper charset attribute first.
+			$charset = $http_equiv_tag->getAttribute( 'charset' );
+			if ( ! $charset ) {
+				// If not, check whether the charset is included with the content type, and use that.
+				$content = $http_equiv_tag->getAttribute( 'content' );
+
+				$matches = [];
+				if ( preg_match( '/;\s*charset=(?<charset>[^;]+)/', $content, $matches ) && ! empty( $matches['charset'] ) ) {
+					$charset = $matches['charset'];
+				}
+			}
+		}
+
+		// Check for HTML 5 charset meta tag. This overrides the HTML 4 charset.
+		$charset_tag = $this->xpath->query( '//meta[ @charset ]' )->item( 0 );
+		if ( $charset_tag ) {
+			$charset_tag = $charset_tag->parentNode->removeChild( $charset_tag );
+			$charset     = $charset_tag->getAttribute( 'charset' );
+		}
+
+		return $charset;
+	}
+
+	/**
+	 * Always ensure that we have an HTML 5 charset meta tag.
+	 *
+	 * The charset defaults to utf-8, which is also what AMP requires.
+	 *
+	 * @param string|false $charset Optional. Charset that was already detected. False if none. Defaults to false.
+	 */
+	protected function ensure_charset_is_present( $charset = false ) {
+		if ( ! empty( $this->meta_tags[ self::TAG_CHARSET ] ) ) {
+			return;
+		}
+
+		$this->meta_tags[ self::TAG_CHARSET ][] = $this->create_charset_element( $charset ?: static::AMP_CHARSET );
+	}
+
+	/**
+	 * Always ensure we have a viewport tag.
+	 *
+	 * The viewport defaults to 'width=device-width', which is the bare minimum that AMP requires.
+	 */
+	protected function ensure_viewport_is_present() {
+		if ( empty( $this->meta_tags[ self::TAG_VIEWPORT ] ) ) {
+			$this->meta_tags[ self::TAG_VIEWPORT ][] = $this->create_viewport_element( static::AMP_VIEWPORT );
+			return;
+		}
+
+		// Ensure we have the 'width=device-width' setting included.
+		$viewport_tag      = $this->meta_tags[ self::TAG_VIEWPORT ][0];
+		$viewport_content  = $viewport_tag->getAttribute( 'content' );
+		$viewport_settings = array_map( 'trim', explode( ',', $viewport_content ) );
+		$width_found       = false;
+
+		foreach ( $viewport_settings as $index => $viewport_setting ) {
+			list( $property, $value ) = array_map( 'trim', explode( '=', $viewport_setting ) );
+			if ( 'width' === $property ) {
+				if ( 'device-width' !== $value ) {
+					$viewport_settings[ $index ] = 'width=device-width';
+				}
+				$width_found = true;
+				break;
+			}
+		}
+
+		if ( ! $width_found ) {
+			array_unshift( $viewport_settings, 'width=device-width' );
+		}
+
+		$viewport_tag->setAttribute( 'content', implode( ',', $viewport_settings ) );
+	}
```

Comment:
This will need to obtain the tag spec for this meta tag essentially incorporate this logic:

https://github.com/ampproject/amp-wp/blob/f15760234e6b2b6bae81769f2e80077fcf6d657d/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php#L1593-L1652

Otherwise, if someone creates a meta tag like:

```html
<meta name=viewport content="width=content-width,BOGUS=SUPER_BOGUS">
```

Then the tag-and-attribute sanitizer will end up _removing_ this `meta` tag that you created here.

In order to facilitate obtaining the tag spec for `meta name=viewport`, instead of iterating over all `meta` tags it and finding the one that has that `spec_name`, it would be useful to perhaps to refactor the spec generation logic to use the `spec_name` as the array key, like so:

```diff
--- a/includes/sanitizers/class-amp-allowed-tags-generated.php
+++ b/includes/sanitizers/class-amp-allowed-tags-generated.php
@@ -10106,7 +10106,7 @@ class AMP_Allowed_Tags_Generated {
 					'unique' => true,
 				),
 			),
-			array(
+			'meta name=viewport' => array(
 				'attr_spec_list' => array(
 					'content' => array(
 						'mandatory' => true,
@@ -10135,7 +10135,6 @@ class AMP_Allowed_Tags_Generated {
 				'tag_spec' => array(
 					'mandatory' => true,
 					'mandatory_parent' => 'head',
-					'spec_name' => 'meta name=viewport',
 					'spec_url' => 'https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml#required-markup',
 					'unique' => true,
 				),
``` 

Then we could extend the `\AMP_Allowed_Tags_Generated::get_allowed_tag()` with a new `$spec_name` arg:

```diff
--- a/includes/sanitizers/class-amp-allowed-tags-generated.php
+++ b/includes/sanitizers/class-amp-allowed-tags-generated.php
@@ -17625,10 +17625,13 @@ class AMP_Allowed_Tags_Generated {
 	 *
 	 * @since 0.7
 	 * @param string $node_name Tag name.
+	 * @param string $spec_name Spec name, to reduce the tag specs to a single one.
 	 * @return array|null Allowed tag, or null if the tag does not exist.
 	 */
-	public static function get_allowed_tag( $node_name ) {
-		if ( isset( self::$allowed_tags[ $node_name ] ) ) {
+	public static function get_allowed_tag( $node_name, $spec_name = null ) {
+		if ( isset( $spec_name, self::$allowed_tags[ $node_name ][ $spec_name ] ) ) {
+			return self::$allowed_tags[ $node_name ][ $spec_name ];
+		} elseif ( isset( self::$allowed_tags[ $node_name ] ) ) {
 			return self::$allowed_tags[ $node_name ];
 		}
 		return null;
```

And then here in the `ensure_viewport_is_present` it could easily locate the tag spec via:

```php
$tag_spec = \AMP_Allowed_Tags_Generated::get_allowed_tag( 'meta', 'meta name=viewport' );
```

This would be useful elsewhere that we refer to tag specs, including in the style sanitizer.

Reply:
More generally, would it be possible to refactor the tag & meta sanitizer to strip the offending properties only, and then check if the entire is still valid or not? This way, we wouldn't have to duplicate this logic all over the place...

> it would be useful to perhaps to refactor the spec generation logic to use the spec_name as the array key

Is this guaranteed not to have collisions?

Reply:
> More generally, would it be possible to refactor the tag & meta sanitizer to strip the offending properties only, and then check if the entire thing is still valid or not? This way, we wouldn't have to duplicate this logic all over the place...

Yes, this could be done. There could be validation errors with type `invalid_meta_property`. I like that.

> Is this guaranteed not to have collisions?

It turns out, yes! Sometimes the `spec_name` is omitted when it is just a regular HTML element, I think. And for `script` we just need to derive a `spec_name` from the `extension_spec`. When this is done, all tag specs are unique: https://gist.github.com/westonruter/8e6a8ca427c69b23cdd8e26dcccbfc3d

Reply:
As discussed during the plugin sync, I'll tackle the enhanced validation-stripping-granularity in a separate PR to keep this one moderately sane.

Reply:
Related: #3780. May be best to wait to work on this until that PR is merged.

Reply:
Actually, I can include the property sanitization as part of #3780.

Reply:
Or rather a subsequent PR built off of that PR.

Reply:
Opened a new issue to track this: #4070